### PR TITLE
Fix failed lotStandingConnection queries when a sale artwork is unpublished

### PR DESCRIPTION
**Pull request by erikdstock**
The Causality LotStanding type needs to be stitched and assembled using data from Gravity. This PR fixes a bug where unpublished gravity data ruins a graphql query, when really it just means that the lot should be treated as nonexistent.

Causality's part is really just the calculated state portion of a gravity sale artwork, and the `saleArtwork` field is stitched into the underlying type in order to make a query like `auctionsLotStandingConnection.edges[0].node.saleArtwork.sale.name`. This works fine unless the sale artwork is unpublished; additionally the sale artwork is implicitly queried for currency-related fields in order to resolve the `display` values. If a sale artwork is unpublished, the entire query fails.

WIP solution: 
- get a list of lot standings (lotStandingConnection) as before but then 
  - fetch the sale artwork for each (currently we have to do this one by one), 
  - catching errors (currently, all errors) and returning `null` for that sale artwork,
  - filter the list for not-nullness,
  - filter the lotStandingConnection `edges` for presence of a matching sale artwork
  - and map the edge to a new edge that includes the sale artwork on its node,
  - return the original connection with our new, filtered + sale-artwork-enhanced edges

If this method works for us we may want to do something similar from the other direction for [PURCHASE-2200]. It would be nice to get some outside buy-in since i'm not sure if this is a new approach or what gotchas might exist.
(mobbed with @lilyfromseattle , @zephraph , @sepans , @starsirius )

[PURCHASE-2200]: https://artsyproduct.atlassian.net/browse/PURCHASE-2200

### CODE REVIEW

File: src/lib/stitching/causality/stitching.ts [hunk 1]
```diff
@@ -118,16 +118,50 @@ export const causalityStitchingEnvironment = ({
...
+                      .saleArtworkRootLoader(lot.internalID)
+                      .catch(() => null)
+                  }
+                )
+
+                const availableSaleArtworks = (
+                  await Promise.all(promisedSaleArtworks)
+                ).filter((sa) => sa !== null)
+
+                const availableEdges = lotStandingsConnection.edges.reduce(
```
Comment (sweir27) [inline src/lib/stitching/causality/stitching.ts:147]: I know this is WIP so I won't comment on the extra logs, etc.!

But overall the approach of querying for lot standings from causality, removing entries where the sale artwork doesn't exist (which we have to determine from gravity) and then returning the results seems 👍.

It is annoying that we have to query one-by-one for the sale artworks 😅. A stop-gap solution could be to add an `ids` param to gravity (I don't think one exists yet) so we can easily batch that request. Definitely not blocking for this, though!

**Comment by erikdstock** [comment]
@sepans and I briefly tried to write a test to validate the new stitching code that adds a `saleArtwork` to the existing `lotStandingConnection` nodes and filters those without one... but we were unable to figure out how to stub out all those tests correctly.

**Comment by sepans** [comment]
I think we were not setting `preserveResolvers` in [`addmockfunctionstoschemaoptions`](https://www.apollographql.com/docs/apollo-server/api/graphql-tools/#addmockfunctionstoschemaoptions) so the resolvers we added were overwritten.

File: src/lib/stitching/causality/__tests__/stitching.test.ts [hunk 2]
```diff
@@ -157,4 +155,136 @@ describe("causality/stitching", () => {
       },
     })
   })
+
+  describe("auctionsLotStandingConnection", () => {
+    it("stitches auctionsLotStandingConnection under Me", async () => {
```
Comment (sepans) [inline src/lib/stitching/causality/__tests__/stitching.test.ts:160]: This test just verifies that stitching is working

File: src/lib/stitching/causality/__tests__/stitching.test.ts [hunk 3]
```diff
@@ -157,4 +155,136 @@ describe("causality/stitching", () => {
...
+                { node: { rawId: "Hello World" } },
+                { node: { rawId: "Hello World" } },
+              ],
+            },
+          },
+        },
+      })
+    })
+
+    describe("when sale artworks are missing", () => {
```
Comment (sepans) [inline src/lib/stitching/causality/__tests__/stitching.test.ts:209]: These two just test the resolver to check if it behaves correctly when `saleArtworkRootLoader` throws error(s). I first tried to combine these but realized separating stitching and resolvers tests make things much simpler. Is there any other edge case that needs to be covered?

File: src/lib/stitching/causality/stitching.ts [hunk 4]
```diff
@@ -118,16 +118,47 @@ export const causalityStitchingEnvironment = ({
...
+                fieldName: "_unused_auctionsLotStandingConnection",
+                args: {
+                  userId: parent.internalID,
+                },
+                context,
+                info,
+              })
+              .then(async (lotStandingsConnection) => {
+                const promisedSaleArtworks = lotStandingsConnection.edges.map(
+                  ({ node: { lot } }) => {
```
Comment (sepans) [inline src/lib/stitching/causality/stitching.ts:134]: One general graphql question: When graphql stitches `_unused_auctionsLotStandingConnection` does it load all the properties? Or does it lazy load things that are being used in the resolver? or it only fetches the fields that the query has asked for?

So in this case specifically is there a guarantee that the `edges->node->lot->internalID` is available even when it is not in the original query? Or we need to add a fragment similar to [this](https://github.com/artsy/metaphysics/pull/2885/files#diff-2a0633fb36f80b8a6471189fc8867efc5686541af5a1be81d4d880932f1edb9aR111) to make sure it is loaded?

**Review by erikdstock — COMMENTED** [review]
Amazing. My only question is whether we want to be more specific about the errors we catch - basically to filter for only known non-visible cases (eg unpublished, HTTP status forbidden i think?)

File: src/lib/stitching/causality/stitching.ts (hunk 4 above)
Comment (sepans) [reply src/lib/stitching/causality/stitching.ts:134]: I will trace the code and verify 👆🏽

Comment (sepans) [reply src/lib/stitching/causality/stitching.ts:134]: This actually the case 😬 without `internalID` under `lot` in the query the resolver doesn't return anything. I will try to see if it is possible to add something in the fragment there to make make sure that is always being fetched...

File: src/lib/stitching/causality/stitching.ts [hunk 5]
```diff
@@ -118,16 +118,47 @@ export const causalityStitchingEnvironment = ({
           // (eg first, after, last, before) are forwarded automatically, so we only
```
Comment (mzikherman) [inline src/lib/stitching/causality/stitching.ts:118]: Is that true? I would expect this to need a 

```
args: {
  userID: parent.internalID,
  ..._args
}
```

to make sure those connection args get passed through.

File: src/lib/stitching/causality/stitching.ts (hunk 1 above)
Comment (mzikherman) [reply src/lib/stitching/causality/stitching.ts:144]: Yea, definite 👍  on adding a better first-class support if there's a need to do a batch check on if a group of artworks/sale artworks are published.

**Review by mzikherman — COMMENTED** [review]
This feels ok, but a tad less than ideal. Why the link between parsing out sale artworks that error from Gravity (which also requires the `internalID` field to be requested in a deeply nested object as @sepans pointed out), from the underlying Causality data? I imagine that Causality lot standing data returns useful information irrespective of any Gravity link, so it feels unexpected that a client _wouldn't_ be able to access those, just cause they error in Gravity. Also, this basically filters after pagination, which I think is better avoided for consistency's sake (a client might receive less items than they requested even if there is a subsequent page, which can be kind of odd).

Overall this resolver I guess sort of feels like an awkward place to do this? I'm wondering if the client (I'm not sure how/where this data is currently used on the front-end), should be able to deal with a sale artwork erroring out from Gravity, or if the Causality GraphQL resolver is the place that this functionality of checking w/ Gravity should go (which would also address the issue about requiring the `internalID` field selection). Perhaps there should be a first-class link between Gravity and Causality on a published state, with a Sidekiq job enqueued in Gravity to update a flag in Causality when an auction artwork is removed (and the Causality API could update to not return those), or whatever.

File: src/lib/stitching/causality/stitching.ts (hunk 5 above)
Comment (starsirius) [reply src/lib/stitching/causality/stitching.ts:118]: I didn't verify, but from the [doc](https://www.graphql-tools.com/docs/schema-delegation/#args-recordstring-any), the arguments are automatically preserved if the subschema is expecting them.

Comment (sepans) [reply src/lib/stitching/causality/stitching.ts:118]: passing `first` to connection works so I _think_ they are implicitly being passed.

**Comment by sepans** [comment]
Thanks @mzikherman for the comments. Very helpful.

> Pagination:

Yes good point. Another sign that this is a less than ideal solution.

> first-class link between Gravity and Causality

I think this is also what @erikdstock had in mind as an ideal solution.

We also considered catching gravity error [here|https://github.com/artsy/metaphysics/blob/6762fbb3947fe51bce0985d30f6a181883725feb/src/lib/stitching/causality/stitching.ts#L66) (and also [here](https://github.com/artsy/metaphysics/blob/6762fbb3947fe51bce0985d30f6a181883725feb/src/lib/stitching/causality/stitching.ts#L12)) and return `null` (or ideally an error type) and leave it to the client to decide about displaying lot standings without `saleArtwork` depending on the context.

In the end, we decided to go with what we have now because it is easy to reverse and we are going to do a fast follow up with a more elegant solution.

**Comment by erikdstock** [comment]
Hey @sepans and @mzikherman, 

I'm wondering if we feel okay merging this as-is with a commitment to come up with a better solution soon.
- Pro: We probably can't really release the new Eigen inbox with My Bids without this fix.
- Con: It relies on some undocumented query shape magic and points to a larger issue in how causality and gravity interact (or fail to)

I also wonder if this will break relay pagination. Currently pagination is not fully supported on the causality relay connection, but say you could request 10, but #10 is not present so you only get 9, and (assuming relay doesn't immediately get confused) when you request everything after cursor #9 you again don't get #10 (now #1), and so on.

In short I think there are risks here and that this will immediately become blocking technical debt, including to folks who ask for more bid history than we initially show. At the same time we have several possible paths to fix it. If this solution will serve us for now I would say let's merge it, test it, and prioritize a more complete fix soon.

**Comment by sepans** [comment]
I don't know enough about the nuances of eigen->metaphyisics dependencies for users who update to the current version but not to the future versions, but I feel this solution doesn't change the schema and also doesn't dramatically change the expectation for this `auctionsLotStandingConnection` type so I prefer merging this to unblock the release and do a better-researched follow-up than to rush a more complex solution to be able to release.

File: src/lib/stitching/causality/stitching.ts (hunk 1 above)
Comment (zephraph) [reply src/lib/stitching/causality/stitching.ts:145]: When we were initially pairing on this we called this out as a definite improvement. This could probably use the batch loader treatment like sales already have.

**Comment by sepans** [comment]
Ok we synched up about this with @zephraph and decided that we can merge this and preferably address the problem upstream where causality returns the lots.
